flink-core/src/main/java/org/apache/flink/api/dag/Transformation.java

```diff
@@ -108,11 +109,10 @@
     public static final int UPPER_BOUND_MAX_PARALLELISM = 1 << 15;
 
     // This is used to assign a unique ID to every Transformation
-    protected static Integer idCounter = 0;
+    protected static final AtomicInteger ID_COUNTER = new AtomicInteger(0);
```

Comment:
nit: does this pr needs a test case with multi threads?

Reply:
I can try to write a test which can try to reproduce the race condition. But it's not always a certainty that this will occur. But anyways, I will try to write a unit test for this.

Reply:
Any update on this @shba24  ?

Reply:
I will finalize this PR.